fence/man/fence_vmware.8
.TH fence_vmware 8

.SH NAME
fence_vmware - I/O Fencing agent for VMware virtual machines

.SH SYNOPSIS
.B
fence_vmware
[\fIOPTION\fR]...

.SH DESCRIPTION
fence_vmware is an I/O Fencing agent which can be used with the VMware ESX,
VMware ESXi or VMware Server to fence virtual machines.

Before you can use this agent, it must be installed VI Perl Toolkit or vmrun
command on every node you want to make fencing.

VI Perl Toolkit is preferred for VMware ESX/ESXi and Virtual Center. Vmrun
command is only solution for VMware Server 1/2 (this command will works against
ESX/ESXi 3.5 up2 and VC up2 too, but not cluster aware!) and is available as part of
VMware VIX API SDK package. VI Perl and VIX API SDK are both available from
VMware web pages (not int RHEL repository!).

You can specify type of VMware you are connecting to with \fB-d\fP switch
(or \fIvmware_type\fR for stdin). Possible values are esx, server2 and server1.
Default value is esx, which will use VI Perl. With server1 and server2, vmrun command
is used.

After you have successfully installed VI Perl Toolkit or VIX API, you should
be able to run fence_vmware_helper (part of this agent) or vmrun command.
This agent supports only vmrun from version 2.0.0 (VIX API 1.6.0).

fence_vmware accepts options on the command line as well as from stdin.
Fenced sends parameters through stdin when it execs the agent. fence_vmware
can be run by itself with command line options. This is useful for testing
and for turning outlets on or off from scripts.

.SH HOST NAME/IP ADDRESS SPECIFICATION
Host name can include port in standard notation host_name:port.
Specifying port isn't mandatory, but you must use it, if you try use this agent
against VMware server 2.0.0 with default options, where https console runs on port
8333. In case of ESX/ESXi, you don't need to specify port, because
default configuration runs https on standard (443) port.

.SH OPTIONS
.TP
\fB-a\fP \fIIPaddress\fR
IP address or hostname of ESX/ESXi host machine. See (see
.SM
.B "HOST NAME/IP ADDRESS SPECIFICATION"
above).
.TP
\fB-h\fP
Print out a help message describing available options, then exit.
.TP
\fB-l\fP \fIlogin\fR
Login name to VMware administration console.
.TP
\fB-o\fP \fIaction\fR
The action required.  This can be reboot (default), status, off, on, list
or monitor.
.TP
\fB-p\fP \fIpassword\fR
Password for login for VMware administration console.
.TP
\fB-S\fP \fIscript\fR
Script to run to retrieve password for VMware administration console.
.TP
\fB-n\fP \fIname\fR
Name of virtual machine to fence (ie. "test" for ESX with VI Perl or
"[datastore1] test/test.vmx" for Server 2) which is returned by list command.
.TP
\fB-e\fP \fIcommand\fR
Location of fence_vmware_helper or vmrun command. Default is fence_vmware_helper or
/usr/bin/vmrun, but if you are using nonstandard location, you may find this switch useful.
.TP
\fB-d\fP \fItype\fR
Type of VMware product you are trying to connect. This can be esx, server1 or server2.
.TP
\fB-s\fP \fIdatacenter\fR
VMware datacenter to use. This can be used to filter guests to operate. Without specifying,
guests from all datacenters will be used. Parameter can be used only with VI Perl helper
(type is esx).
.TP
\fB-v\fP
Verbose. Record session to stdout, or debug file if specified (see -D).
.TP
\fB-D\fP
Specifies file, where will be written debug messages from session.
.TP
\fB-V\fP
Print out a version message, then exit.

.SH STDIN PARAMETERS
.TP
\fIagent = < param >\fR
This option is used by fence_node(8) and is ignored by fence_vmware.
.TP
\fIipaddr = < hostname | ip >\fR
IP address or hostname of ESX/ESXi host machine. See (see
.SM
.B "HOST NAME/IP ADDRESS SPECIFICATION"
above).
.TP
\fIlogin = < param >\fR
Login name to VMware administration console.
.TP
\fIoption = < param >\fR
The action required.  This can be reboot (default), status, off, on, list
or monitor.
.TP
\fIpasswd = < param >\fR
Password for login for VMware administration console.
.TP
\fIpasswd_script = < param >\fR
Script to run to retrieve password for VMware administration console.
.TP
\fIport = < param >\fR
Name of virtual machine to fence (ie. "test" for ESX with VI Perl or
"[datastore1] test/test.vmx" for Server 2) which is returned by list command.
.TP
\fIexec = < param >\fR
Location of fence_vmware_helper or vmrun command. Default is fence_vmware_helper or
/usr/bin/vmrun, but if you are using nonstandard location, you may find this switch useful.
.TP
\fIvmware_type = < param >\fR
Type of VMware product you are trying to connect. This can be esx, server1 or server2.
.TP
\fIvmware_datacenter = < param >\fR
VMware datacenter to use. This can be used to filter guests to operate. Without specifying,
guests from all datacenters will be used. Parameter can be used only with VI Perl helper
(type is esx).
.TP
\fIverbose = < param >\fR
Verbose.  Record session to stdout, or debug file if specified (see debug).
.TP
\fIdebug = < param >\fR
Specifies file, where will be written debug messages from session.

.SH SEE ALSO
fence(8), fence_node(8)


fence/agents/vmware/fence_vmware.py
#!/usr/bin/python

#
# The Following agent has been tested on:
# vmrun 2.0.0 build-116503 (from VMware Server 2.0) against:
# 	VMware ESX 3.5 (works correctly)
# 	VMware Server 2.0.0 (works correctly)
#	VMware ESXi 3.5 update 2 (works correctly)
# 	VMware Server 1.0.7 (works but list/status show only running VMs)
#
# VI Perl API 1.6 against:
# 	VMware ESX 3.5
#	VMware ESXi 3.5 update 2
# 	VMware Virtual Center 2.5
#

import sys, re, pexpect
sys.path.append("@FENCEAGENTSLIBDIR@")
from fencing import *

#BEGIN_VERSION_GENERATION
RELEASE_VERSION="VMware Agent using VI Perl API and/or VIX vmrun command"
REDHAT_COPYRIGHT=""
BUILD_DATE=""
#END_VERSION_GENERATION

### CONSTANTS ####
# VMware type is ESX/ESXi/VC
VMWARE_TYPE_ESX=0
# VMware type is Server 1.x
VMWARE_TYPE_SERVER1=1
# VMware type is Server 2.x and/or ESX 3.5 up2, ESXi 3.5 up2, VC 2.5 up2
VMWARE_TYPE_SERVER2=2

# Minimum required version of vmrun command
VMRUN_MINIMUM_REQUIRED_VERSION=2

# Default path to vmhelper command
VMHELPER_COMMAND="fence_vmware_helper"
# Default path to vmrun command
VMRUN_COMMAND="/usr/bin/vmrun"
# Default type of vmware
VMWARE_DEFAULT_TYPE="esx"

#### GLOBAL VARIABLES ####
# Internal type. One of VMWARE_TYPE_, set by #vmware_check_vmware_type
vmware_internal_type=VMWARE_TYPE_ESX

# If ESX is disconnected, say, that VM is off (don't return previous state)
vmware_disconnected_hack=False

### FUNCTIONS ####

#Split string in simplified DSV format to array of items
def dsv_split(dsv_str):
	delimiter_c=':'
	escape_c='\\'

	res=[]
	status=0
	tmp_str=""

	for x in dsv_str:
		if (status==0):
			if (x==delimiter_c):
				res.append(tmp_str)
				tmp_str=""
			elif (x==escape_c):
				status=1
			else:
				tmp_str+=x
		elif (status==1):
			if (x==delimiter_c):
				tmp_str+=delimiter_c
			elif (x==escape_c):
				tmp_str+=escape_c
			else:
				tmp_str+=escape_c+x
			status=0

	if (tmp_str!=""):
		res.append(tmp_str)

	return res

# Return string with command and additional parameters (something like vmrun -h 'host'
def vmware_prepare_command(options,add_login_params,additional_params):
	res=options["-e"]

	if (add_login_params):
		if (vmware_internal_type==VMWARE_TYPE_ESX):
			res+=" --server '%s' --username '%s' --password '%s' "%(options["-a"],options["-l"],options["-p"])
		elif (vmware_internal_type==VMWARE_TYPE_SERVER2):
			res+=" -h 'https://%s/sdk' -u '%s' -p '%s' -T server "%(options["-a"],options["-l"],options["-p"])
		elif (vmware_internal_type==VMWARE_TYPE_SERVER1):
			host_name_array=options["-a"].split(':')

			res+=" -h '%s' -u '%s' -p '%s' -T server1 "%(host_name_array[0],options["-l"],options["-p"])
			if (len(host_name_array)>1):
				res+="-P '%s' "%(host_name_array[1])

	if ((options.has_key("-s")) and (vmware_internal_type==VMWARE_TYPE_ESX)):
		res+="--datacenter '%s' "%(options["-s"])

	if (additional_params!=""):
		res+=additional_params

	return res

# Log message if user set verbose option
def vmware_log(options, message):
	if options["log"] >= LOG_MODE_VERBOSE:
		options["debug_fh"].write(message+"\n")

# Run command with timeout and parameters. Internaly uses vmware_prepare_command. Returns string
# with output from vmrun command. If something fails (command not found, exit code is not 0), fail_usage
# function is called (and never return).
def vmware_run_command(options,add_login_params,additional_params,additional_timeout):
	command=vmware_prepare_command(options,add_login_params,additional_params)

	try:
		vmware_log(options,command)

		(res_output,res_code)=pexpect.run(command,SHELL_TIMEOUT+LOGIN_TIMEOUT+additional_timeout,True)

		if (res_code==None):
			fail(EC_TIMED_OUT)
		if ((res_code!=0) and (add_login_params)):
			vmware_log(options,res_output)
			fail_usage("%s returned %s"%(options["-e"],res_output))
		else:
			vmware_log(options,res_output)

	except pexpect.ExceptionPexpect:
		fail_usage("Cannot run command %s"%(options["-e"]))

	return res_output

# Get outlet list with status as hash table. If you will use add_vm_name, only VM with vmname is
# returned. This is used in get_status function
def vmware_get_outlets_vi(conn, options, add_vm_name):
	outlets={}

	if (add_vm_name):
		all_machines=vmware_run_command(options,True,("--operation status --vmname '%s'"%(options["-n"])),0)
	else:
		all_machines=vmware_run_command(options,True,"--operation list",POWER_TIMEOUT)

	all_machines_array=all_machines.splitlines()

	for machine in all_machines_array:
		machine_array=dsv_split(machine)
		if (len(machine_array)==4):
			if (machine_array[0] in outlets):
				fail_usage("Failed. More machines with same name %s found!"%(machine_array[0]))

			if (vmware_disconnected_hack):
				outlets[machine_array[0]]=("",(
						((machine_array[2].lower() in ["poweredon"]) and
						 (machine_array[3].lower()=="connected"))
						and "on" or "off"))
			else:
				outlets[machine_array[0]]=("",((machine_array[2].lower() in ["poweredon"]) and "on" or "off"))
	return outlets

# Get outlet list with status as hash table.
def vmware_get_outlets_vix(conn,options):
	outlets={}

	running_machines=vmware_run_command(options,True,"list",0)
	running_machines_array=running_machines.splitlines()[1:]

	if (vmware_internal_type==VMWARE_TYPE_SERVER2):
		all_machines=vmware_run_command(options,True,"listRegisteredVM",0)
		all_machines_array=all_machines.splitlines()[1:]
	elif (vmware_internal_type==VMWARE_TYPE_SERVER1):
		all_machines_array=running_machines_array

	for machine in all_machines_array:
		if (machine!=""):
			outlets[machine]=("",((machine in running_machines_array) and "on" or "off"))

	return outlets

def get_outlets_status(conn, options):
	if (vmware_internal_type==VMWARE_TYPE_ESX):
		return vmware_get_outlets_vi(conn,options,False)
	if ((vmware_internal_type==VMWARE_TYPE_SERVER1) or (vmware_internal_type==VMWARE_TYPE_SERVER2)):
		return vmware_get_outlets_vix(conn,options)

def get_power_status(conn,options):
	if (vmware_internal_type==VMWARE_TYPE_ESX):
		outlets=vmware_get_outlets_vi(conn,options,True)
	else:
		outlets=get_outlets_status(conn,options,False)

	if ((vmware_internal_type==VMWARE_TYPE_SERVER2) or (vmware_internal_type==VMWARE_TYPE_ESX)):
		if (not (options["-n"] in outlets)):
			fail_usage("Failed: You have to enter existing name of virtual machine!")
		else:
			return outlets[options["-n"]][1]
	elif (vmware_internal_type==VMWARE_TYPE_SERVER1):
		return ((options["-n"] in outlets) and "on" or "off")

def set_power_status(conn, options):
	if (vmware_internal_type==VMWARE_TYPE_ESX):
		additional_params="--operation %s --vmname '%s'"%((options["-o"]=="on" and "on" or "off"),options["-n"])
	elif ((vmware_internal_type==VMWARE_TYPE_SERVER1) or (vmware_internal_type==VMWARE_TYPE_SERVER2)):
		additional_params="%s '%s'"%((options["-o"]=="on" and "start" or "stop"),options["-n"])
		if (options["-o"]=="off"):
			additional_params+=" hard"

	vmware_run_command(options,True,additional_params,POWER_TIMEOUT)

# Returns True, if user uses supported vmrun version (currently >=2.0.0) otherwise False.
def vmware_is_supported_vmrun_version(options):
	vmware_help_str=vmware_run_command(options,False,"",0)
	version_re=re.search("vmrun version (\d\.(\d[\.]*)*)",vmware_help_str.lower())
	if (version_re==None):
		    return False   # Looks like this "vmrun" is not real vmrun

	version_array=version_re.group(1).split(".")

	try:
		if (int(version_array[0])<VMRUN_MINIMUM_REQUIRED_VERSION):
			return False
	except Exception:
		return False

	return True

# Define new options
def vmware_define_defaults():
	all_opt["vmware_type"]["default"]=VMWARE_DEFAULT_TYPE

# Check vmware type, set vmware_internal_type to one of VMWARE_TYPE_ value and
# options["-e"] to path (if not specified)
def vmware_check_vmware_type(options):
	global vmware_internal_type

	options["-d"]=options["-d"].lower()

	if (options["-d"]=="esx"):
		vmware_internal_type=VMWARE_TYPE_ESX
		if (not options.has_key("-e")):
			options["-e"]=VMHELPER_COMMAND
	elif (options["-d"]=="server2"):
		vmware_internal_type=VMWARE_TYPE_SERVER2
		if (not options.has_key("-e")):
			options["-e"]=VMRUN_COMMAND
	elif (options["-d"]=="server1"):
		vmware_internal_type=VMWARE_TYPE_SERVER1
		if (not options.has_key("-e")):
			options["-e"]=VMRUN_COMMAND
	else:
		fail_usage("vmware_type can be esx,server2 or server1!")

# Main agent method
def main():
	device_opt = [ "help", "version", "agent", "quiet", "verbose", "debug",
		       "action", "ipaddr", "login", "passwd", "passwd_script",
		       "test", "port", "separator", "exec", "vmware_type",
		       "vmware_datacenter", "secure" ]

	vmware_define_defaults()

	options = check_input(device_opt, process_input(device_opt))

	# Default is secure connection
	options["-x"] = 1

	# Check vmware type and set path
	vmware_check_vmware_type(options)

	# Test user vmrun command version
	if ((vmware_internal_type==VMWARE_TYPE_SERVER1) or (vmware_internal_type==VMWARE_TYPE_SERVER2)):
		if (not (vmware_is_supported_vmrun_version(options))):
			fail_usage("Unsupported version of vmrun command! You must use at least version %d!"%(VMRUN_MINIMUM_REQUIRED_VERSION))

	# Operate the fencing device
	fence_action(None, options, set_power_status, get_power_status, get_outlets_status)

if __name__ == "__main__":
	main()
